text/0000-signing-registry-commits.md

```diff
@@ -58,41 +58,247 @@ index is updated if the registry in question is a registry that ought to be
+and `expires` fields.
+
+The `timestamp.toml` file consists of a toplevel object with the following
+members:
+
+- `spec-version`: Version number of the TUF specification (`1` for now).
+- `version`: Counter (starting at `1`) which is incremented when
+  `timestamp.toml` is updated.
+- `expires`: date (as described in [RFC 3339][rfc3339]) which determines
+  when the index should be considered stale and no longer trusted by clients.
+
+See also subsection "4.6. File formats: timestamp.json" in section
+[4. Document formats][tuf-spec-s4] in [The Update Framework Specification][tuf-spec].
+Note that the `"meta"` key from `timestamp.json` is omitted in this
+implementation. See "TUF Implementation Notes" following this section.
+
+#### Example `timestamp.toml`
+
+```toml
+spec-version = "1"
+version = "123"
+expires = "2019-01-01T00:00:00Z"
+```
+
+### TUF Implementation Notes
+
+The implementation of TUF described above uses the concepts, metadata files,
+schema, and structures from the upstream project, but diverges from the
+upstream implementation in the following ways:
+
+- **Git + OpenPGP signatures**: git's content hashing and OpenPGP signatures
+  are used to authenticate all files. This provides what TUF describes as
+  "consistent snapshots", i.e. all files are modified in atomic commits and
+  clients always see a linear history. However, this has some overlap with
+  TUF's signature formats and certain TUF roles (most notably the "snapshot"
+  role, see below)
+- **TOML**: used as an alternative serialization format vs TUF's  CJSON. This
+  shouldn't be too problematic as TUF is already designed to support multiple
+  serialization formats (including e.g. ASN.1 DER), and existing Rust
+  implementations such as `rust-tuf` are already designed to support multiple
+  serialization formats. Furthermore, Rust serialization libraries like serde
+  make supporting multiple serialization formats simple.
+- **OpenPGP V4 key fingerprints**: used as TUF key IDs, domain separated from
+  TUF's native key ID format with the `openpgp:` prefix. This is done for
+  two reasons: to avoid having two different key ID formats, and because we
+  aren't using CJSON and therefore can't implement TUF's CJSON-based key
+  fingerprinting algorithm. Instead, the key IDs are ordinary OpenPGP v4 key
+  fingerprints which are computed using the SHA-1 hash function (see notes on
+  SHA-1 in the "Security Considerations" section at the end of this document),
+  and hash a message that includes an algorithm identifier and the public key
+  (see items `d` and `e` in section section [12.2][rfc4880-s12.2] of
+  [RFC 4880][rfc4480])
+- **`timestamp` and `root` roles combined**: this RFC describes a setup where
+  in practice all `root` keys would also be authorized as `timestamp` keys,
+  which allows us to enforce the property that all commits are signed by the
+  `timestamp` role. There is no security advantage to separating them (at least
```

Comment:
If the timestamp key is used to ensure "transparency" (i.e., to prevent the creation of 2 separate commit chains), then we have to require an "old timestamp" signature on all commits.

Reply:
In general preventing duplicity in systems like this is difficult (see similar concerns regarding CT logs, auditors, gossip protocols, etc).

That said, I don't follow what you're saying. A signature over a git commit covers the current `tree` in addition to the `parent` OID, which is the SHA-1 digest of the previous commit's `tree`, `parent`, and the full OpenPGP signature over that commit (provided it's there), i.e. each commit signature is over the entire repository history including the signature on the previous commit.

Reply:
The idea is to have the timestamp key is stored in a manner that makes it hard for an attacker to create unlogged signatures (I think Amazon offers an HSM module that does that). If we can find such a service, it would be worth to put the timestamp key in it.

Reply:
Using cloud HSM or KMS services on any of the major cloud vendors (e.g. AWS, GCP, Azure) would likely require using a different signature algorithm than Ed25519, e.g. ECDSA over P-256. ECDSA has wide support among cloud KMS services, HSMs, and PIV devices such as YubiKey 4s.

That said, there are several hardware options for Ed25519 key storage. YubiHSM2s support Ed25519, as do a number of cryptocurrency hardware wallets (e.g. Ledger), although whether that's a practical deployment option for something like @bors seems questionable.

Reply:
Sure, but I'll like the scheme to be ready for secure deployment with an HSM.

Reply:
I think ECDSA support is worth considering. FWIW I support it in [Signatory](https://docs.rs/signatory/0.8.0/signatory/ecdsa/index.html).